### PR TITLE
[yang]Add FG-ECMP yang model

### PR DESCRIPTION
**Pull request by lmingz**
Why I did it
Add yang model definition for FG_NHG_MEMBER, FG_NHG and FG_NHG_PREFIX from request: https://github.com/sonic-net/sonic-buildimage/issues/12457

How I did it
Add yang model refereeing to description https://github.com/sonic-net/SONiC/blob/master/doc/ecmp/fine_grained_next_hop_hld.md

How to verify it
pyang tests

Which release branch to backport (provide reason below if selected)
 201811
 201911
 202006
 202012
 202106
 202111
 202205
[x ] 202211
Description for the changelog
Add yang model for FG_NHG_MEMBER, FG_NHG and FG_NHG_PREFIX

### CODE REVIEW

**Comment by lmingz** [comment]
Hi @ganglyu and @anish-n, please review this pr.

**Review by ganglyu — APPROVED** [review]

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang [hunk 1]
```diff
@@ -0,0 +1,171 @@
+module sonic-fine-grained-ecmp {
+
+    yang-version 1.1;
+
+    namespace "http://github.com/Azure/sonic-fine-grained-ecmp";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang:5]: Please use sonic-net to replace Azure

Comment (lmingz) [reply src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang]: switched to sonic-net

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                        "next_hop_ip": "10.1.0.1",
+                        "FG_NHG": "group2",
+                        "bank": 3,
+                        "link": "Ethernet1"
+                    }
+                ]
+            }
+        }
+    },
+    "FG_NHG_TEST_DUPLICATE_NAME": {
```
Comment (anish-n) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:106]: Whats the difference between this one and FG_NHG_TEST_DUPLICATE_ENTRY, looks pretty similar to me as a test case

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                        "next_hop_ip": "10.0.0.1",
+                        "FG_NHG": "group1",
+                        "bank": 1,
+                        "link": "Ethernet1"
+                    },
+                    {
+                        "next_hop_ip": "10.0.0.2",
+                        "FG_NHG": "group1",
+                        "bank": 2,
+                        "link": "Ethernet1"
```
Comment (anish-n) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:94]: Can we add a test case where an invalid port is provided as the "link"?

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                    }
+                ]
+            },
+            "sonic-fine-grained-ecmp:FG_NHG_MEMBER": {
+                "FG_NHG_MEMBER_LIST": [
+                    {
+                        "next_hop_ip": "10.0.0.1",
+                        "FG_NHG": "group1",
+                        "bank": 1,
+                        "link": "Ethernet1"
```
Comment (anish-n) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:88]: Can we add a test case with member 1 pointing to link 1, and then member 2 pointing to link 2.

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+                leaf ip_prefix{
+                    type stypes:sonic-ip-prefix;
+                    description "FG_NHG_PREFIX for which FG behavior is
+                    desired";
+                }
+
+                leaf FG_NHG{
+                    type leafref {
+						path "/sfgecmp:sonic-fine-grained-ecmp/sfgecmp:FG_NHG/sfgecmp:FG_NHG_LIST/sfgecmp:name";
```
Comment (anish-n) [inline src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang:112]: identation issue?

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+                leaf next_hop_ip{
+                    type inet:ip-address;
+                    description "FG_NHG next-hop-ip member associated
+                    with prefix";
+                }
+
+                leaf FG_NHG{
+                    type leafref {
+						path "/sfgecmp:sonic-fine-grained-ecmp/sfgecmp:FG_NHG/sfgecmp:FG_NHG_LIST/sfgecmp:name";
```
Comment (anish-n) [inline src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang:138]: Identation issue?

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json (hunk 2 above)
Comment (lmingz) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:106]: removed

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json (hunk 3 above)
Comment (lmingz) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:94]: Invalid link test case added.

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang (hunk 5 above)
Comment (lmingz) [reply src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang:112]: resolved

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang (hunk 6 above)
Comment (lmingz) [reply src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang:138]: resolved

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json (hunk 4 above)
Comment (lmingz) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:88]: Added a multiple port, multiple member test case.

File: src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang (hunk 6 above)
Comment (lmingz) [reply src/sonic-yang-models/yang-models/sonic-fine-grained-ecmp.yang]: fixed

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+                        "name": "group2",
+                        "bucket_size": 20,
+                        "match_mode": "route-based"
+                    }
+                ]
+            },
+            "sonic-fine-grained-ecmp:FG_NHG_PREFIX": {
+                "FG_NHG_PREFIX_LIST": [
+                    {
+                        "ip_prefix": "10.0.0.0/24",
```
Comment (anish-n) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:39]: Add test case for Ipv6 prefix as well

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+                        "name": "group2",
+                        "bucket_size": 20,
+                        "match_mode": "route-based"
+                    }
+                ]
+            },
+            "sonic-fine-grained-ecmp:FG_NHG_MEMBER": {
+                "FG_NHG_MEMBER_LIST": [
+                    {
+                        "next_hop_ip": "10.0.0.1",
```
Comment (anish-n) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:85]: Add test case for IPv6 next-hop-ip as well

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json (hunk 7 above)
Comment (lmingz) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:39]: ipv6 cases added

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json (hunk 8 above)
Comment (lmingz) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/fine_grained_ecmp.json:93]: ipv6 cases added.

**Comment by lmingz** [comment]
@zhangyanzhao please review this new pr about FG_ECMP yang model.

**Review by ganglyu — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Review by anish-n — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
@lguohan  @yxieca  can you please help to merge this PR? Thanks.

**Pull request merged**